File: Group 6 Project /smart_campus/lib/models/attendance.dart
class Attendance {
  final int id;
  final int studentId;
  final String? studentName;
  final int courseId;
  final String? courseName;
  final String date;
  final String status;
  final int? markedBy;

  Attendance({
    required this.id,
    required this.studentId,
    this.studentName,
    required this.courseId,
    this.courseName,
    required this.date,
    required this.status,
    this.markedBy,
  });

  factory Attendance.fromJson(Map<String, dynamic> json) {
    return Attendance(
      id: json['id'] is int ? json['id'] : (int.tryParse(json['id'].toString()) ?? 0),
      studentId: json['studentId'] ?? json['student_id'] ?? 0,
      studentName: json['studentName'] ?? json['student_name'],
      courseId: json['courseId'] ?? json['course_id'] ?? 0,
      courseName: json['courseName'] ?? json['course_name'],
      date: json['date'] ?? '',
      status: json['status'] ?? 'present',
      markedBy: json['markedBy'] ?? json['marked_by'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'studentId': studentId,
      'studentName': studentName,
      'courseId': courseId,
      'courseName': courseName,
      'date': date,
      'status': status,
      'markedBy': markedBy,
    };
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'studentId': studentId,
      'studentName': studentName,
      'courseId': courseId,
      'courseName': courseName,
      'date': date,
      'status': status,
      'markedBy': markedBy,
    };
  }

  factory Attendance.fromMap(Map<String, dynamic> map) {
    return Attendance(
      id: map['id'] is int ? map['id'] : (int.tryParse(map['id'].toString()) ?? 0),
      studentId: map['studentId'] ?? map['student_id'] ?? 0,
      studentName: map['studentName'] ?? map['student_name'],
      courseId: map['courseId'] ?? map['course_id'] ?? 0,
      courseName: map['courseName'] ?? map['course_name'],
      date: map['date'] ?? '',
      status: map['status'] ?? 'present',
      markedBy: map['markedBy'] ?? map['marked_by'],
    );
  }
}

File: Group 6 Project /backend/routes/attendance.js
const express = require('express');
const db = require('../config/db');
const auth = require('../middleware/auth');
const authorize = require('../middleware/role');

const router = express.Router();
router.use(auth);

router.post('/', authorize('faculty'), (req, res) => {
    const course_id = req.body.course_id ?? req.body.courseId;
    const date = req.body.date;
    const records = req.body.records || []; // records: [{student_id or studentId, status}]
    try {
        const deleteExisting = db.prepare('DELETE FROM attendance WHERE course_id = ? AND date = ?');
        const insert = db.prepare('INSERT INTO attendance (student_id, course_id, date, status, marked_by) VALUES (?, ?, ?, ?, ?)');
        const insertMany = db.transaction((recordsToInsert) => {
            deleteExisting.run(course_id, date);
            for (const record of recordsToInsert) {
                const studentId = record.student_id ?? record.studentId;
                insert.run(studentId, course_id, date, record.status, req.user.id);
            }
        });
        insertMany(records);
        res.status(201).json({ message: 'Attendance marked successfully' });
    } catch (error) {
        res.status(400).json({ error: error.message });
    }
});

router.get('/course/:courseId', authorize('admin', 'faculty'), (req, res) => {
    const { date } = req.query;
    const { courseId } = req.params;
    try {
        let attendance;
        if (date) {
            attendance = db.prepare(`
                SELECT a.*, u.name as student_name 
                FROM attendance a 
                JOIN users u ON a.student_id = u.id 
                WHERE a.course_id = ? AND a.date = ?
            `).all(courseId, date);
        } else {
            attendance = db.prepare(`
                SELECT a.*, u.name as student_name 
                FROM attendance a 
                JOIN users u ON a.student_id = u.id 
                WHERE a.course_id = ?
            `).all(courseId);
        }
        res.json(attendance);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

router.get('/student/:studentId', (req, res) => {
    const { studentId } = req.params;
    
    // Check if student is accessing their own attendance or if it's admin/faculty
    if (req.user.role === 'student' && req.user.id !== parseInt(studentId)) {
        return res.status(403).json({ error: 'Forbidden: Can only access own attendance' });
    }

    try {
        const attendance = db.prepare(`
            SELECT a.*, c.name as course_name, u.name as student_name 
            FROM attendance a 
            JOIN courses c ON a.course_id = c.id 
            JOIN users u ON a.student_id = u.id 
            WHERE a.student_id = ?
        `).all(studentId);
        res.json(attendance);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

module.exports = router;
